kernel/comps/keyboard/src/i8042_chip/controller.rs

```diff
@@ -0,0 +1,282 @@
+        &mut self,
+        command: Command,
+    ) -> Result<(), I8042ControllerError> {
+        for _ in 0..MAX_WAITING_COUNT {
+            if self.send_command(command).is_ok() {
+                return Ok(());
+            }
+            core::hint::spin_loop();
+        }
+        Err(I8042ControllerError::DeviceBusy)
+    }
+
+    pub(super) fn send_command(&mut self, command: Command) -> Result<(), I8042ControllerError> {
+        if !self.read_status().contains(Status::INPUT_BUFFER_IS_FULL) {
+            self.write_command(command as u8);
+            Ok(())
+        } else {
+            Err(I8042ControllerError::DeviceBusy)
+        }
+    }
+
+    pub(super) fn wait_and_send_data(&mut self, data: u8) -> Result<(), I8042ControllerError> {
+        for _ in 0..MAX_WAITING_COUNT {
+            if self.send_data(data).is_ok() {
+                return Ok(());
+            }
+            core::hint::spin_loop();
+        }
+        Err(I8042ControllerError::DeviceBusy)
+    }
+
+    pub(super) fn send_data(&mut self, data: u8) -> Result<(), I8042ControllerError> {
+        if !self.read_status().contains(Status::INPUT_BUFFER_IS_FULL) {
+            self.write_data(data);
+            Ok(())
+        } else {
+            Err(I8042ControllerError::DeviceBusy)
+        }
+    }
+
+    pub(super) fn wait_and_recv_data(&mut self) -> Result<u8, I8042ControllerError> {
+        for _ in 0..MAX_WAITING_COUNT {
+            if let Some(data) = self.receive_data() {
+                return Ok(data);
+            }
+            core::hint::spin_loop();
+        }
+        Err(I8042ControllerError::DeviceNoInput)
+    }
+
+    pub(super) fn receive_data(&mut self) -> Option<u8> {
+        if self.read_status().contains(Status::OUTPUT_BUFFER_IS_FULL) {
+            Some(self.read_data())
+        } else {
+            None
+        }
+    }
+
+    fn read_configuration(&mut self) -> Result<Configuration, I8042ControllerError> {
+        self.wait_and_send_command(Command::ReadConfiguration)?;
+        Ok(Configuration::from_bits_retain(self.wait_and_recv_data()?))
+    }
+
+    fn write_configuration(&mut self, config: Configuration) -> Result<(), I8042ControllerError> {
+        self.wait_and_send_command(Command::WriteConfiguration)?;
+        self.wait_and_send_data(config.bits())
+    }
+
+    fn read_data(&self) -> u8 {
+        self.data_port.read()
+    }
+
+    fn write_data(&mut self, data: u8) {
+        self.data_port.write(data);
+    }
+
+    fn read_status(&self) -> Status {
+        Status::from_bits_retain(self.status_or_command_port.read())
+    }
+
+    fn write_command(&mut self, command: u8) {
+        self.status_or_command_port.write(command);
+    }
+}
+
+/// Errors that can occur when initializing the i8042 controller.
+#[derive(Debug, Clone, Copy)]
+pub(super) enum I8042ControllerError {
+    ControllerTestFailed,
+    FirstPortTestFailed,
+    SecondPortTestFailed,
+    DeviceBusy,
+    DeviceNoInput,
```

Comment:
`DeviceBusy` -> `OutputBusy`, `DeviceNoInput` -> `NoInput` as these errors may also occur when sending the controller command or receiving the controller response.